cabal-install/Distribution/Client/CmdRepl.hs

```diff
@@ -100,62 +212,223 @@ replCommand = Client.installCommand {
+    finalizer
   where
     verbosity = fromFlagOrDefault normal (configVerbosity configFlags)
     cliConfig = commandLineFlagsToProjectConfig
                   globalFlags configFlags configExFlags
                   installFlags haddockFlags
 
+withProject :: ProjectConfig -> Verbosity -> [String] -> IO (ProjectBaseContext, [TargetSelector], IO ())
+withProject cliConfig verbosity targetStrings = do
+  baseCtx <- establishProjectBaseContext verbosity cliConfig
+
+  targetSelectors <- either (reportTargetSelectorProblems verbosity) return
+                 =<< readTargetSelectors (localPackages baseCtx) targetStrings
+
+  return (baseCtx, targetSelectors, return ())
+
+withoutProject :: ProjectConfig -> Verbosity -> [String]  -> IO (ProjectBaseContext, [TargetSelector], IO ())
+withoutProject config verbosity extraArgs = do
+  unless (null extraArgs) $
+    die' verbosity $ "'repl' doesn't take any extra arguments when outside a project: " ++ unwords extraArgs
+
+  globalTmp <- getTemporaryDirectory
+  tempDir <- createTempDirectory globalTmp "cabal-repl."
+
+  baseCtx <- establishDummyProjectBaseContext
+              verbosity
+              config
+              tempDir
+              []
+  
+  pkgDb <- projectConfigWithBuilderRepoContext 
+            verbosity
+            (buildSettings baseCtx)
+            (getSourcePackages verbosity)
+
+  -- We need to create a dummy package that lives in our dummy project.
+  let
+    (basePkg:_) = sortOn (Down . packageId) $ 
+      PackageIndex.lookupPackageName (packageIndex pkgDb) "base"
+
+    sourcePackage = SourcePackage
+      { packageInfoId        = pkgId
+      , packageDescription   = genericPackageDescription
+      , packageSource        = LocalUnpackedPackage tempDir
+      , packageDescrOverride = Nothing
+      }
+    genericPackageDescription = emptyGenericPackageDescription 
+      & L.packageDescription .~ packageDescription
+      & L.condLibrary        .~ Just (CondNode library [] [])
+    packageDescription = emptyPackageDescription
+      { package = pkgId
+      , specVersionRaw = Left (mkVersion [2, 2])
+      , library = Just library
+      , licenseRaw = Left SPDX.NONE
+      }
+    library = emptyLibrary { libBuildInfo = buildInfo }
+    buildInfo = emptyBuildInfo
+      { targetBuildDepends = [pkgIdToDependency (packageId basePkg)]
+      , defaultLanguage = Just Haskell2010
+      }
+    pkgId = PackageIdentifier "fake-package" version0
+
+  putStrLn $ showGenericPackageDescription genericPackageDescription
+  writeGenericPackageDescription (tempDir </> "fake-package.cabal") genericPackageDescription
+
+  baseCtx' <- establishDummyProjectBaseContext
+            verbosity
+            config
+            tempDir
+            [SpecificSourcePackage sourcePackage]
+
+  let
+    targetSelectors = [TargetPackage TargetExplicitNamed [pkgId] Nothing]
+    finalizer = return () --handleDoesNotExist () (removeDirectoryRecursive tempDir)
```

Comment:
Hmm, so `finalizer` is `return ()` in both cases?

Reply:
Nope, that's just for debugging. This is a WIP.